### PR TITLE
Add LEAF node for AVL trees (half memory usage)

### PR DESCRIPTION
**Pull request by sjoelund**

### CODE REVIEW

**Comment by OpenModelica-Hudson** [comment]
The job failed to compile; for details, see [OpenModelica_TEST_PULL_REQUEST 2016-04-21_14-33-21](https://test.openmodelica.org/hudson//job/OpenModelica_TEST_PULL_REQUEST/2513/console).

File: Compiler/Util/BaseAvlSet.mo [hunk 1]
```diff
@@ -134,6 +156,7 @@ algorithm
   key := match inTree
     case EMPTY() algorithm return; then fail();
     case NODE() then inTree.key;
+    case LEAF() then inTree.key;
```
Comment (hkiel) [inline Compiler/Util/BaseAvlSet.mo]: better do `else inTree.key;` instead of two cases NODE() and LEAF()?

Comment (sjoelund) [inline Compiler/Util/BaseAvlSet.mo]: That would be a compile-time error (because we do not know if it is NODE or LEAF

Comment (hkiel) [inline Compiler/Util/BaseAvlSet.mo]: Then in (new) line 167 add `else false;` to account for the new record NODE(), otherwise the match fails.

Comment (sjoelund) [inline Compiler/Util/BaseAvlSet.mo]: True. It should not fail...

**Comment by OpenModelica-Hudson** [comment]
The job failed to compile; for details, see [OpenModelica_TEST_PULL_REQUEST 2016-04-21_17-10-54](https://test.openmodelica.org/hudson//job/OpenModelica_TEST_PULL_REQUEST/2514/console).

**Comment by hkiel** [comment]
applying the patch locally works fine... (did not try bootstrapping)

**Comment by hkiel** [comment]
Any try to make this work?
My calculations assume less than half the memory usage (balanced binary tree has ~50% leaves and 50% inner nodes, so for half of the total nodes you use 2 instead of 5 words, makes a reduction of 30% in memory)

**Comment by sjoelund** [comment]
Working on MODRIO deliverables. I will try to make it work. Note that the number of words used by a uniontype is 2+n where n is the number of fields. So 3 vs. 6 for sets, 4 vs. 7 for trees. However, GC allocates in chunks of certain sizes (8/16 bytes; I guess 2 words on 32/64-bit platforms, not sure) so it might be actually using 4 vs. 6 for sets and 4 vs. 8 for trees.

For example, GC reports 104.6 GB memory allocations while we only requested a total of 83695461544  bytes (77.9 GB) of allocations. I suspect a lot of the memory allocations are lists, which are 3 words (with an overhead ~33%).

**Comment by sjoelund** [comment]
The Tree.get did not have a case for LEAF

**Comment by OpenModelica-Hudson** [comment]
The test suite is unstable according to [OpenModelica_TEST_PULL_REQUEST 2016-04-26_11-05-39](https://test.openmodelica.org/hudson//job/OpenModelica_TEST_PULL_REQUEST/2538/).

**Comment by sjoelund** [comment]
And a problem with the rotate functions. Also, when rotating, NODE was always created even when it was rotated into a LEAF.

**Comment by OpenModelica-Hudson** [comment]
The tests run correctly according to [OpenModelica_TEST_PULL_REQUEST 2016-04-26_12-38-33](https://test.openmodelica.org/hudson//job/OpenModelica_TEST_PULL_REQUEST/2540/console).

**Pull request closed without merging**